grpc/src/main/java/io/stargate/grpc/service/Service.java

```diff
@@ -77,12 +84,18 @@
 import org.apache.cassandra.stargate.exceptions.WriteFailureException;
 import org.apache.cassandra.stargate.exceptions.WriteTimeoutException;
 import org.immutables.value.Value;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class Service extends io.stargate.proto.StargateGrpc.StargateImplBase {
-
+  private static final Logger log = LoggerFactory.getLogger(Service.class);
   public static final Context.Key<AuthenticationSubject> AUTHENTICATION_KEY =
       Context.key("authentication");
   public static final Context.Key<SocketAddress> REMOTE_ADDRESS_KEY = Context.key("remoteAddress");
+  private static final String SYSTEM_TRACES_KEYSPACE = "system_traces";
+  public static final String TRACING_PREPARE_QUERY =
+      "select activity, source, source_elapsed, thread from events where session_id = ?";
```

Comment:
I'm pretty sure you need to query the  `system_traces.session` table first. I think it acts as "gate" to signifying tracing is complete.  I'm digging into the server-side implementation more to verify.

Check the implementation in the java-driver: https://github.com/datastax/java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/internal/core/cql/QueryTraceFetcher.java

It queries "SELECT * FROM system_traces.sessions WHERE session_id = ?" and validates `duration` and `started_`:

`row == null || row.isNull("duration") || row.isNull("started_at")`

Reply:
minor: would prefer: `SELECT activity, source, source_elapsed, thread FROM events WHERE session_id = ?`. Makes this easier to see in code as a query.

Reply:
thank you for this, I've implemented a new `QueryTracingFetcher` that uses a similar mechanism to java driver: firstly queries the `session` table then `events`. 
I've moved the tracing fetcher to the `persistence-api` module. The client of this class needs to map `Rows` to the expected type ourselves - for `gRPC`, see the `TraceEventsMapper`.
I didn't move it directly to the `DataStore` because of: https://github.com/stargate/stargate/pull/1077#discussion_r662044829